pkg/restore/actions/csi/volumesnapshot_action.go

```diff
@@ -188,13 +143,85 @@ func (p *volumeSnapshotRestoreItemAction) Progress(
 	operationID string,
 	restore *velerov1api.Restore,
 ) (velero.OperationProgress, error) {
-	return velero.OperationProgress{}, nil
+	progress := velero.OperationProgress{}
+
+	// The operationID format should be "vsNamespace/vsName".
+	if operationID == "" || !strings.Contains(operationID, "/") {
+		return progress, riav2.InvalidOperationIDError(operationID)
+	}
+
+	logger := p.log.WithFields(logrus.Fields{
+		"Action":      "VolumeSnapshotRestoreItemAction",
+		"OperationID": operationID,
+		"Namespace":   restore.Namespace,
+	})
+
+	vsc := new(snapshotv1api.VolumeSnapshotContent)
+	vs := new(snapshotv1api.VolumeSnapshot)
+
+	operationIDParts := strings.Split(operationID, "/")
+
+	if err := p.crClient.Get(
+		context.TODO(),
+		crclient.ObjectKey{Namespace: operationIDParts[0], Name: operationIDParts[1]},
+		vs,
+	); err != nil {
+		logger.Errorf("Fail to get the VolumeSnapshot %s", operationID)
+		return progress, err
+	}
+
+	if vs.Spec.Source.VolumeSnapshotContentName == nil {
+		logger.Errorf("VolumeSnapshot %s referenced VSC is nil", operationID)
+		return progress, errors.Errorf("VolumeSnapshot %s referenced VSC is nil", operationID)
+	}
+
+	if err := p.crClient.Get(
+		context.TODO(),
+		crclient.ObjectKey{
+			Name: *vs.Spec.Source.VolumeSnapshotContentName,
+		},
+		vsc,
+	); err != nil {
+		logger.Errorf("Fail to get the VolumeSnapshotContent %s",
+			*vs.Spec.Source.VolumeSnapshotContentName,
+		)
+		return progress, err
+	}
+
+	if (vsc.Status != nil && boolptr.IsSetToTrue(vsc.Status.ReadyToUse)) &&
+		(vs.Status != nil && boolptr.IsSetToTrue(vs.Status.ReadyToUse)) {
+		// VS and VSC are ready. Delete them to make sure they will not cause conflict for the next restore.
+		if err := p.crClient.Delete(context.TODO(), vs); err != nil {
+			logger.Warnf("Fail to delete VS %s after VS turns ready %s.", vs.Namespace+vs.Name, err.Error())
+		}
+		if err := p.crClient.Delete(context.TODO(), vsc); err != nil {
+			logger.Warnf("Fail to delete VSC %s after VSC turns ready %s.", vsc.Name, err.Error())
+		}
+	} else {
+		logger.Debugf("VolumeSnapshot or VolumeSnapshotContent are not ready yet.")
+		progress.Description = "VolumeSnapshot or VolumeSnapshotContent creation are still in progress."
+		return progress, nil
+	}
+
+	progress.Description = "VolumeSnapshot and VolumeSnapshotContent are ready"
+	progress.OperationUnits = "Bytes"
+	progress.Completed = true
+	if vsc.Status.RestoreSize != nil {
```

Comment:
Don't set `progress.NCompleted` and `progress.NTotal` as discussed previously

Reply:
There should not be a progress function according to design.
I will modify it.